src/contextmenu.js

```diff
@@ -0,0 +1,472 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Multiple selection context menu.
+ */
+
+import * as Blockly from 'blockly/core';
+import {blockSelection, hasSelectedParent} from './global';
+
+/**
+ * Modification for context menu 'Duplicate' to be available for
+ * multiple blocks.
+ */
+const registerDuplicate = function() {
+  const duplicateOption = {
+    displayText: function() {
+      const selectedBlocksLength = blockSelection.size;
+      if (selectedBlocksLength <= 1) {
+        return Blockly.Msg['DUPLICATE_BLOCK'];
+      } else {
+        return Blockly.Msg['DUPLICATE_BLOCK'] + ' (' +
+        selectedBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      const block = scope.block;
+      if (!block.isInFlyout && block.isDeletable() && block.isMovable()) {
+        if (block.isDuplicatable()) {
+          return 'enabled';
+        }
+        return 'disabled';
+      }
+      return 'hidden';
+    },
+    callback: function(scope) {
+      const duplicatedBlocks = [];
+      const apply = function(block) {
+        if (block && duplicateOption.preconditionFn({
+          block: block,
+        }) === 'enabled' && !hasSelectedParent(block)) {
```

Comment:
```suggestion
        if (block &&
            duplicateOption.preconditionFn({block: block}) === 'enabled' &&
            !hasSelectedParent(block)) {
```

Nit: formatting. Generally best to include line breaks at a higher syntactic level, because it makes the code easier to read.

Here and below.

Reply:
I think you can also remove the check for `block` because you always check that it exists before calling `apply`. Here and below.